socket/sc_sock.c

```diff
@@ -1724,21 +1941,140 @@ uint32_t sc_sock_poll_event(struct sc_sock_poll *p, int i)
 	return evs;
 }
 
+static int sc_sock_poll_fill_results(struct sc_sock_poll *p)
+{
+	int found = 0;
+
+	for (int i = p->results_offset; i < p->cap && p->results_remaining > 0; i++) {
+		enum sc_sock_ev events = sc_sock_poll_event_inner(p, i);
+
+		if (events != SC_SOCK_NONE) {
+			p->results[found++] = (struct sc_sock_poll_result){
+				.events = events,
+				.data = sc_sock_poll_data_inner(p, i)->data,
+			};
+
+			if (found == SC_SOCK_POLL_MAX_EVENTS) {
+				p->results_offset = i + 1;
+				return found;
+			}
+		}
+	}
+
+	// We only can reach this line when either p->results_remaining is
+	// already 0 or when i reaches p->cap, in the latter case we need to
+	// reset p->results_remaining to 0 because we are missing some expected
+	// events after processing submitted socket ops.
+	p->results_remaining = 0;
+	return found;
+}
+
 int sc_sock_poll_wait(struct sc_sock_poll *p, int timeout)
 {
-	int n, rc = p->cap;
+	int n, rc = 0;
 
 	timeout = (timeout == -1) ? 16 : timeout;
 
+	if (p->events == NULL) {
+		sc_sock_poll_set_err(p, "poll : sc_sock_poll is not initialized or already terminated");
+		return -1;
+	}
+
+	EnterCriticalSection(&p->lock);
+	if (p->polling) {
+		rc = -1;
+		sc_sock_poll_set_err(p, "poll : polling is already in progress in a parallel thread");
+		goto exit;
+	}
+	if (p->results_remaining > 0) {
+		assert(p->results_offset >= 1);
+		// Fill the remaining results that did not fit into p->results
+		// in the previous iteration.
+		rc = sc_sock_poll_fill_results(p);
+		goto exit;
+	}
+
+	// When p->polling is set to true, no add/del operations can happen,
+	// instead these operations will be submitted to the sc_sock_poll->ops
+	// and processed by the poller thread after the polling is finished.
+	p->polling = true;
+	LeaveCriticalSection(&p->lock);
+
 	do {
+		// When p->polling is set to true p->events and p->cap are
+		// stable and can be read outside of the lock.
 		n = WSAPoll(p->events, (ULONG) p->cap, timeout);
 	} while (n < 0 && errno == EINTR);
 
-	if (n == SC_INVALID) {
+	if (n > 0) {
+		// sc_sock_poll_event_inner() will decrement
+		// p->results_remaining when we have some events found.
+		p->results_remaining = n;
+
+		// Drain wakeup_pipe only when we can do that without blocking.
+		if (sc_sock_poll_event_inner(p, 0) == SC_SOCK_READ) {
+			// To reduce contention we drain wakeup_pipe outside of
+			// the lock. It is safe to do that before the processing
+			// of p->ops, this way we do not drain any extra wakeup
+			// signals.
+			//
+			// At this point we may have any number of bytes in
+			// wakeup_pipe, we just pick reasonably large buffer
+			// size for draining and assume that sc_sock_pipe_read()
+			// supports partial reads
+			// (less than the provided buffer size).
+			char buf[16];
+			sc_sock_pipe_read(&p->wakeup_pipe, buf, sizeof(buf));
+		}
+	} else {
+		// Because otherwise we would not even start polling.
+		assert(p->results_remaining == 0);
```

Comment:
I think we'll hit here if `n == SOCKET_ERROR`. Should we move `n == SOCKET_ERROR` if check above?

Reply:
We will also hit here if either `n == SOCKET_ERROR` or `n == 0`, in both cases we should expect `results_remaining` to be `0`. I will add assert and comment for that.